### PR TITLE
Feature/1841 currency

### LINKED ISSUE #1841: Input type text add new property "currency"
It was requested, that we allow automatic separators and rounding for input fields.

For instance: "10250.123" => "CHF 10'250.15"

We will support the INTL Format api, available native in the browser (https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat)

Currency attribute should receive currency value like so: `currency="CHF"`. Locale can be fixed to `de-CH`.

### PR DESCRIPTION
**Pull request by danyball**
Fixes #1841

# Merge Checklist:
- [x] Code selfreview has been performed ([Best Practices](https://github.com/axa-ch/patterns-library/blob/develop/CONTRIBUTION.md#best-practices)).
- [x] Tests are sufficient.
- [x] Modifications still work in IE.
- [x] Documentation is up to date.
- [x] Changelog is up to date.
- [x] Typescript typings for properties are up to date.
- [x] Designers approved changes or no approval needed.

- [x] is popup change ok? --> #1969

### CODE REVIEW

**Comment by danyball** [comment]
Open issues:

- [x] chars left wrong

- [x] currency input: user type "hello" > what should happen?

File: src/components/20-molecules/popup/index.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ export * from './popup-mixin';
 /* eslint-disable no-undef */
 const axaPopupVersion = __VERSION_INFO__;
 /* eslint-enable no-undef */
-const commonVersion = axaPopupVersion['axa-popup']['axa-popup'];
+const commonVersion = axaPopupVersion['axa-popup'];
```
Comment (danyball) [inline src/components/20-molecules/popup/index.js]: This has nothing to do with the currency feature directly. But without that change the Unit Tests are failing. @markus-walther can you please check that this change is ok?

**Comment by danyball** [comment]
Tests laufen alle lokal durch

>  408 passed (7m 30s)

File: src/components/10-atoms/input-text/index.js [hunk 2]
```diff
@@ -290,6 +319,12 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
       _open,
     } = this;
 
+    let formattedValue = value;
+    if (document.activeElement !== this.nativeInput) {
+      // Do not format the value if the user is still typing. We will format its input onBlur.
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/index.js]: onBlur => on blur. (Comments should be in plain English, onBlur is not a word)

File: src/components/10-atoms/input-text/ui.test.js [hunk 3]
```diff
@@ -209,3 +209,18 @@ test('should display correct default value', async t => {
...
+test('should format value', async t => {
+  const $axaInputElement = await Selector(() => document.querySelector(TAG), {
+    dependencies: { TAG },
+  }).find(CLASS);
+
+  await t.typeText($axaInputElement, '1234', { replace: true });
+  await t.pressKey('tab'); // to blur input-element
+
+  await t.expect($axaInputElement.value).eql('CHF 1’234.00');
+});
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/ui.test.js:226]: I am missing a controlled-mode test. For that it should be defined what controlled mode means when currency is active...

**Review by markus-walther — CHANGES_REQUESTED** [review]
Found some small things, otherwise good work!

Will check later why the popup issue you find was not causing problems before...

File: src/components/10-atoms/input-text/index.js [hunk 4]
```diff
@@ -248,6 +252,31 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
...
+    if (currency && type === 'text' && !this.currencyFormatter) {
+      // just create a new Intl if it does not exist
+      this.currencyFormatter = new Intl.NumberFormat('de-CH', {
+        style: 'currency',
+        currency,
+      });
+    }
+
+    if (this.currencyFormatter) {
+      const containsNumber = /\d/g;
```
Comment (raphaellueckl) [inline src/components/10-atoms/input-text/index.js]: does this regex check if a value HAS a number, or if it ONLY is a number (and no letters?). because then the name is a bit misleading.

File: src/components/10-atoms/input-text/index.js [hunk 5]
```diff
@@ -248,6 +252,31 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
...
+
+    if (currency && type === 'text' && !this.currencyFormatter) {
+      // just create a new Intl if it does not exist
+      this.currencyFormatter = new Intl.NumberFormat('de-CH', {
+        style: 'currency',
+        currency,
+      });
+    }
+
+    if (this.currencyFormatter) {
```
Comment (raphaellueckl) [inline src/components/10-atoms/input-text/index.js]: i don't really like the structure of those if-else blocks in here.

line 266 checks if currencyformatter is set, which is an obsolete check if the type is "number" (which is already checked previously). this is not a clean approach.

better would be (pseudocode):

```
_formatCurrency(value) {
  if (currency && type=="number") { // if not type currency or number, don't look any further...
    if (!this.currencyFormatter) {
       // INIT
    }
    // proceed the formatting...
  }
  return value
}
`
``

**Review by raphaellueckl — CHANGES_REQUESTED** [review]

File: src/components/10-atoms/input-text/index.js (hunk 4 above)
Comment (danyball) [reply src/components/10-atoms/input-text/index.js]: HAS. Because only if input contains a number the formatter can work correctly

File: src/components/10-atoms/input-text/ui.test.js (hunk 3 above)
Comment (danyball) [reply src/components/10-atoms/input-text/ui.test.js:226]: done.

File: src/components/10-atoms/input-text/index.js (hunk 4 above)
Comment (01231) [reply src/components/10-atoms/input-text/index.js]: what about `const isNumber`?

Comment (danyball) [reply src/components/10-atoms/input-text/index.js]: Do not change the variable, name is perfect.

File: src/components/10-atoms/input-text/index.js (hunk 5 above)
Comment (danyball) [reply src/components/10-atoms/input-text/index.js]: done

**Review by raphaellueckl — DISMISSED** [review]

File: src/components/10-atoms/input-text/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,7 @@
+## 4.4.0
+
+- You can now set `currency` attribute. This will format the input `1000` for example to `CHF 1’000.00`. For details see README.md.
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/CHANGELOG.md]: wording: A `currency` attribute

File: src/components/10-atoms/input-text/README.md [hunk 7]
```diff
@@ -160,6 +160,14 @@ It is a string value. Useful to open up numeric keyboard on touch devices. See o
 
 The Boolean attribute autofocus, when true, sets keyboard focus on the underlying native &lt;input&gt; element after initial rendering of the component.
 
+### currency
+
+You can set this string value to format numbers. It formats the input when the `onBlur` event of the input element is fired or if you set `value` programmatically. You can set `ISO 4217` [currency codes](https://en.wikipedia.org/wiki/ISO_4217), but the locale "style" is fixed to `de-CH`.
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/README.md]: Wording:

You can set the `currency` attribute to a valid [ISO 4217 currency code](https://en.wikipedia.org/wiki/ISO_4217) string. When set, &lt;input-text&gt; formats the input whenever the `onBlur` event of its internal input element  fires, and also upon setting `value` programmatically.  Note that the locale "style" remains fixed to `de-CH`.

File: src/components/10-atoms/input-text/README.md [hunk 8]
```diff
@@ -160,6 +160,14 @@ It is a string value. Useful to open up numeric keyboard on touch devices. See o
...
+
+The formatter is just working if attribute `type` is set to `text`.
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/README.md]: Wording: Currency formatting is only active when attribute ....

File: src/components/10-atoms/input-text/README.md [hunk 9]
```diff
@@ -160,6 +160,14 @@ It is a string value. Useful to open up numeric keyboard on touch devices. See o
...
+
+Please be aware of using this attribute in combination with `maxLength`. The result of formatting could exceed the limits of `maxLength`!
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/README.md]: Wording: Please be mindful of using currency formatting in combination...

File: src/components/10-atoms/input-text/index.js [hunk 10]
```diff
@@ -248,6 +252,31 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
     this.nativeInput = this.querySelector('input');
   }
 
+  _formatCurrency(value) {
+    const { currency, type } = this;
+    const containsNumber = /\d/g;
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/index.js]: Remove the `g` after the regexp, it serves no purpose here.

The semantics is: has at least 1 digit somewhere, you might want to rename the variable to express this better.

File: src/components/10-atoms/input-text/index.js [hunk 11]
```diff
@@ -248,6 +252,31 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
...
+    if (currency && type === 'text') {
+      if (!this.currencyFormatter) {
+        // just create a new Intl if it does not exist
+        this.currencyFormatter = new Intl.NumberFormat('de-CH', {
+          style: 'currency',
+          currency,
+        });
+      }
+
+      if (containsNumber.test(value)) {
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/index.js]: containsNumber is not reused, so in better DRY style you'd write

const hasAtLeastOneDigit = /\d/.test(value) above and here only if(hasAtLeastOneDigit) { ...}

File: src/components/10-atoms/input-text/index.js [hunk 12]
```diff
@@ -248,6 +252,31 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
...
+        this.invalid = false;
+        const valueDecimalsOnly = value.replace(/[^0-9.]/g, '');
+        return this.currencyFormatter.format(valueDecimalsOnly);
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/index.js]: This means that `mickeymouse19newyork.foobar34` is a valid input... is it good UX to be this lenient?

File: src/components/10-atoms/input-text/index.js [hunk 13]
```diff
@@ -248,6 +252,31 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
...
   }
 
+  _formatCurrency(value) {
+    const { currency, type } = this;
+    const containsNumber = /\d/g;
+
+    if (currency && type === 'text') {
+      if (!this.currencyFormatter) {
+        // just create a new Intl if it does not exist
+        this.currencyFormatter = new Intl.NumberFormat('de-CH', {
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/index.js:263]: Why does the currency formatter need to be part of the element's state?

File: src/components/10-atoms/input-text/index.js [hunk 14]
```diff
@@ -290,6 +319,12 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
       _open,
     } = this;
 
+    let formattedValue = value;
+    if (document.activeElement !== this.nativeInput) {
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/index.js:329]: comment missing: // focus not on internal input?

**Review by markus-walther — CHANGES_REQUESTED** [review]
Good job! Some small findings, please address.

File: src/components/20-molecules/popup/index.js (hunk 1 above)
Comment (markus-walther) [reply src/components/20-molecules/popup/index.js]: This change is definitely NOT ok. It leads to definitions like &lt;axa-popup-button-objectObject&gt; ...
Please revert it and adapt the unit tests instead.

File: src/components/10-atoms/input-text/index.js (hunk 13 above)
Comment (danyball) [reply src/components/10-atoms/input-text/index.js:263]: It is not needed. But i dont find a way to unit test `currencyFormatter` if its a variable of the module.

File: src/components/10-atoms/input-text/index.js (hunk 12 above)
Comment (danyball) [reply src/components/10-atoms/input-text/index.js]: Will ask thomas, but thats ok for me.

File: src/components/10-atoms/input-text/index.js (hunk 14 above)
Comment (danyball) [reply src/components/10-atoms/input-text/index.js:329]: The comment is at line 325? Or what do you mean?

File: src/components/10-atoms/input-text/index.js (hunk 12 above)
Comment (danyball) [reply src/components/10-atoms/input-text/index.js]: It is ok for Thomas.

File: src/components/20-molecules/popup/index.js (hunk 1 above)
Comment (danyball) [reply src/components/20-molecules/popup/index.js]: Reverted and fixed.

File: src/components/10-atoms/input-text/index.js (hunk 14 above)
Comment (01231) [reply src/components/10-atoms/input-text/index.js:329]: you could move the comment one line up

File: src/components/10-atoms/input-text/index.js (hunk 11 above)
Comment (01231) [reply src/components/10-atoms/input-text/index.js]: Good input. Can you implement this?

**Review by 01231 — DISMISSED** [review]
Nice

File: src/components/10-atoms/input-text/index.js [hunk 15]
```diff
@@ -248,6 +252,33 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
...
+        this.currencyFormatter = new Intl.NumberFormat('de-CH', {
+          style: 'currency',
+          currency,
+        });
+      }
+
+      if (hasAtLeastOneDigit && dotsCounted <= 1) {
+        this.invalid = false;
+        const valueDecimalsOnly = value.replace(/[^0-9.]/g, '');
+        return this.currencyFormatter.format(valueDecimalsOnly);
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/index.js]: This is a brittle way to solve the NaN bug that arises with a specific input that has multiple dots. There could be many other patterns.

You need to address the root of the problem with `parseFloat(valueDecimalsOnly)`. If it  parses as a valid floating-point number, it could be used as a currency value.

File: src/components/10-atoms/input-text/index.js (hunk 14 above)
Comment (markus-walther) [reply src/components/10-atoms/input-text/index.js:329]: What I mean is that it is super-unobvious what `if (document.activeElement !== this.nativeInput) ` does without the comment I suggested, placed above or at the end of line 325.

Principle: Any code that requires some deep thinking needs a comment that explains the Why.

This line definitely requires deep thinking or even googling if you didn't do it recently...

**Review by markus-walther — DISMISSED** [review]
Have found two things only this time! After that, it should be fine.

File: src/components/10-atoms/input-text/index.js [hunk 16]
```diff
@@ -267,10 +266,14 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
...
-        this.invalid = false;
+      if (hasAtLeastOneDigit) {
         const valueDecimalsOnly = value.replace(/[^0-9.]/g, '');
-        return this.currencyFormatter.format(valueDecimalsOnly);
+
+        // eslint-disable-next-line no-restricted-globals
+        if (!isNaN(valueDecimalsOnly)) {
+          this.invalid = false;
+          return this.currencyFormatter.format(valueDecimalsOnly);
+        }
```
Comment (01231) [inline src/components/10-atoms/input-text/index.js]: Would use parseFloat(valueDecimalsOnly) as @markus-walther recommended. It is cleaner and more speaking.

**Review by 01231 — CHANGES_REQUESTED** [review]

File: src/components/10-atoms/input-text/index.js (hunk 16 above)
Comment (01231) [reply src/components/10-atoms/input-text/index.js]:
```
if (hasAtLeastOneDigit) {
  const valueDecimalsOnly = value.replace(/[^0-9.]/g, '');
  return this.currencyFormatter.format(parseFloat(valueDecimalsOnly));
 }
```

File: src/components/10-atoms/input-text/README.md [hunk 17]
```diff
@@ -164,7 +164,7 @@ The Boolean attribute autofocus, when true, sets keyboard focus on the underlyin
 
 You can set the `currency` attribute to a valid [ISO 4217 currency code](https://en.wikipedia.org/wiki/ISO_4217) string. When set, &lt;input-text&gt; formats the input whenever the `onBlur` event of its internal input element fires, and also upon setting `value` programmatically. Note that the locale "style" remains fixed to `de-CH`.
 
-Currency formatting is only active when attribute `type` is set to `text`. If the inputted text can not be formatted the attribute `invalid` switches to `true`.
+Currency formatting is only active when attribute `type` is set to `text`. If the inputted text can not be formatted the attribute `invalid` switches to `true`. A valid input to format is for example `12.20`, invalid is `12.20.1` because of the two dots.
```
Comment (01231) [inline src/components/10-atoms/input-text/README.md]: would write that letters only are also invalid. i wouldn't write that 12.20 is valid because this makes the impression that only this format is allowed. Better just write what isn't allowed (to dots and only letters). there are too many allowed input formats.

**Review by 01231 — CHANGES_REQUESTED** [review]
